elm_for_nmmo/elm_helper.py

```diff
@@ -0,0 +1,264 @@
+    # pylint: disable=cell-var-from-loop
+    test_task = make_team_tasks(teams, [single_spec])
+    env.reset(make_task_fn=lambda: test_task)
+    def run_env():
+      for _ in range(3):
+        env.step({})
+      sys.exit(0) # success
+
+    # sometimes the task fn has a long loop, so we need to time out
+    proc = mp.Process(target=run_env)
+    proc.start()
+    start_time = time.time()
+    while proc.is_alive():
+      elapsed_time = time.time() - start_time
+      if elapsed_time > TIME_OUT:
+        print("NMMO task timed out")
+        proc.terminate()
+        break
+      time.sleep(0.1)
+
+    if proc.exitcode == 0:
+      num_success += 1
+
+  return num_success > 0 # at least 1 spec runs
+
+def generate_task_spec(result_str, fn_name, num_sample=3):
+  # pylint: disable=bare-except,exec-used,bad-builtin
+  task_spec = []
+  task_fn_str = extract_task_fn(result_str, fn_name)
+  import_str = "from nmmo.task.game_state import GameState\n" +\
+                "from nmmo.task.group import Group\n" +\
+                "from nmmo.task.base_predicates import *\n\n"
+
+  locals_dict = {}
+  try:
+    # NOTE: this is a security vulenerability
+    # TODO: make this secure
+    exec(import_str + task_fn_str, globals(), locals_dict)
+  except:
+    # return empty task spec for invalid function
+    print('Invalid python function generated ...')
+    return task_spec
+  task_fn = locals_dict[fn_name]
+  fn_params = inspect.signature(task_fn).parameters
+
+  included_kwargs = set()
+  for _ in range(num_sample):
+    eval_kwargs = {}
+    for key, param in fn_params.items():
+      if key in ['gs', 'subject']:
+        continue
+      type_hint = param.annotation.__name__
+      eval_kwargs[key] = sample_parameter(key, type_hint)
+    args_vals = tuple(eval_kwargs.values())
+    if args_vals not in included_kwargs:
+      task_spec.append(('agent', task_fn, eval_kwargs))
+      included_kwargs.add(args_vals)
+
+  return task_spec
+
+def task_spec_to_str(task_spec):
+  # extract task_fn source code from task_spec
+  extracted_fn_list = set()
+  task_fn_src = []
+  for single_spec in task_spec:
+    fn_name = single_spec[1].__name__
+    if fn_name not in extracted_fn_list:
+      task_fn_src.append(inspect.getsource(single_spec[1]))
+      extracted_fn_list.add(fn_name)
+  return "\n".join(task_fn_src)
+
+# get the list of pre-built task functions
+def is_function_type(obj):
+  return inspect.isfunction(obj) and not inspect.isbuiltin(obj)
+
+def extract_module_fn(module: ModuleType):
+  fn_dict = {}
+  for name, fn in module.__dict__.items():
+    if is_function_type(fn) and not name.startswith('_'):
+      fn_dict[name] = fn
+  return fn_dict
+
+PREBUILT_TASK_FN = extract_module_fn(nmmo.task.base_predicates)
+
+# used in OpenELMTaskGenerator: see self.config.env.impr = import_str["short_import"]
+import_str = {"short_import": """from predicates import TickGE,StayAlive,AllDead,EatFood,DrinkWater,CanSeeTile,CanSeeAgent,OccupyTile
```

Comment:
What's all this?

Reply:
These are info that are fed into LLM when running the ELM. We may or may not need to do so, and the elm-for-nmmo has a huge room for improvements.